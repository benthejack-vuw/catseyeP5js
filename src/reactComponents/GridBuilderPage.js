import React from 'react'

const GridBuilderPage = () => (
    <div>
        GRID BUILDER    
    </div>
)

export default GridBuilderPage;