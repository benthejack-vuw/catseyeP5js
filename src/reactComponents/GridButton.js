import React from 'react'
import { renderGrid, patternFill } from '../helperFunctions/gridRenderer'

export default class GridButton extends React.Component{
    
    _canvasRef;

    constructor(props){
        super(props);
        this._canvasRef = React.createRef();
    }

    componentDidMount = ()=>{
        const tile = renderGrid(true, false, this.props);
        this._canvasRef.current.getContext('2d').clearRect(0,0,this.props.width, this.props.height);
        patternFill(tile, this._canvasRef.current);
    }

    render = ()=>(
                <canvas
                    className={
                        this.props.selected
                        ? 'grid-button grid-button--selected'
                        : 'grid-button'
                    }
                    ref={this._canvasRef}
                    width={this.props.width}
                    height={this.props.height}
                    onClick={this.props.onClick}
                />
        )
    

}

