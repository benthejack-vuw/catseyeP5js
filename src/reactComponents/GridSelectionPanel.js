import React from 'react'

import store from '../store/configureStore'


const GridSelectionPanel = (props) => (
    <div>
        {
            props.grids &&
            props.grids.map((grid)=>(
                <GridButton key={grid.name} {...grid} />
            ))
        }
    </div>
)

export default GridSelectionPanel;