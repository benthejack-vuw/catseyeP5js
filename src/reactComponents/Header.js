import React from 'react';
import { Link } from 'react-router-dom'

export const Header = ({startLogout}) => (  
    <div className='header'>
        <div className='container'>
            <header className='header__layout'>
                <Link className="header__home-link" to="/">
                    <h1>Catseye</h1>
                </Link>
                
                <Link className="header__help-link" to="/help">
                    <h2>Help</h2>
                </Link>
            </header>
        </div>
    </div>
);



export default Header;