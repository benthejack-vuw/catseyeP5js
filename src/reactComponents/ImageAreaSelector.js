import React from 'react'

export default class ImageAreaSelector extends React.Component{

    _svgContainerRef;
    _svgImageRef;

	constructor(props){
        super(props);

        this.state = {
            size:props.size,
            coords: [{x:0, y:0}, {x:props.size.x, y:0}, {x:props.size.x, y:props.size.y}],
            selected: undefined
        }

        this._svgContainerRef = React.createRef();
        this._svgImageRef = React.createRef();

        document.addEventListener('mouseup', this.stopDragging);
    }

    render = () => ( 
        <div>
            <svg 
             ref={this._svgContainerRef}
             width={this.state.size.x}
             height={this.state.size.y}
             onMouseMove={this.updateMouseCoords}
            >
                <image 
                    ref={this._svgImageRef}
                    xlinkHref={this.props.image && this.props.image.src}
                    width='100%'
                    onLoad={
                        (() => {
                            this.centerImage()
                            this.constrainSelectorsToImage()
                            this.setFromTextureCoordinates(this.props.textureCoordinates)
                        })
                    }
                />

                {this.dashedLines()}
                {this.selectionCircles()}
            </svg>
        </div>
    )

    componentDidMount = () => {
        this.centerImage();
        this.constrainSelectorsToImage();
    }

    componentWillUnmount = () => {
        document.removeEventListener('mouseup', this.stopDragging);
    }

    updateMouseCoords = (e) => {
        const rect = this._svgContainerRef.current.getBoundingClientRect();
        const x = e.clientX - rect.left; 
        const y = e.clientY - rect.top;

        this.setState((prev)=>{
            const coords = [...prev.coords];
            if(prev.selected !== undefined){
                coords[prev.selected] = {x,y};
            }
            return {coords};
        });

        this.constrainSelectorsToImage();
        this.props.updateCallback(this.uvCoordinates());
    }

    setFromTextureCoordinates = (textureCoordinates) => {
        const bounds = this.imageBounds();

        const coords = textureCoordinates.map((coord)=>{
            return {
                x:coord.x*bounds.width + bounds.left,
                y:coord.y*bounds.height + bounds.top
            }
        });

        this.setState(()=>({ coords }))
    }

    uvCoordinates = () => {
        const bounds = this.imageBounds();

        return this.state.coords.map((coord)=>{
            return {
                x:(coord.x - bounds.left)/bounds.width,
                y:(coord.y - bounds.top)/bounds.height
            }
        });
    }
    
    imageBounds = () => {
        const containerRect = this._svgContainerRef.current.getBoundingClientRect();
        const imageRect = this._svgImageRef.current.getBoundingClientRect();
        return new DOMRect(imageRect.x - containerRect.x, imageRect.y - containerRect.y, imageRect.width, imageRect.height);
    }

    centerImage = () => {
        const imageElem = this._svgImageRef.current;
        const imageRect = imageElem.getBoundingClientRect();
        imageElem.setAttribute('x', (this.state.size.x-imageRect.width)/2 );
        imageElem.setAttribute('y', (this.state.size.y-imageRect.height)/2 );
    }

    constrainSelectorsToImage = () => {
        const bounds = this.imageBounds();
        this.setState(({coords}) => ({
            coords: coords.map(({x, y}) => ({
                x: Math.min(Math.max(bounds.left, x), bounds.right),
                y: Math.min(Math.max(bounds.top, y), bounds.bottom),
            }))
        }))
    }

    idPrefix = () => this.props.id || 'image-area-selector'

    startDragging = (index) => {
        this.setState( ({selected:prevSelected}) => ({selected: !prevSelected ? index : prevSelected}) ) 
    }

    stopDragging = () => {
        this.setState( () => ({selected: undefined}) )
    }

    dashedLines = () => {
        return this.state.coords.map((pt, index, array) => (
            <line 
                key={`selectionLine${index}`}
                strokeDasharray="5, 5"
                x1={pt.x}
                y1={pt.y}
                x2={array[(index+1)%array.length].x}
                y2={array[(index+1)%array.length].y}
                stroke='black'
            />
        ))
    }

    selectionCircles = () => {
        return this.state.coords.map((pt, index) => (
            <circle 
                key={`selection${index}`}
                className='image-selection-circle'                            
                cx={pt.x}
                cy={pt.y}
                r={10}
                onMouseDown = {()=>{this.startDragging(index)}}
            />
        ))
    }

}
