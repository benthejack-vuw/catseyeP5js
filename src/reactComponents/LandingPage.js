import React from 'react'
import { Link } from 'react-router-dom'

const LandingPage = () => (
    <div>

        <Link to='/pattern-builder/'>
            <p>Build a pattern</p>
        </Link>

        <Link to='/grid-builder/'>
            <p>Build a grid</p>
        </Link>
    
    </div>
)

export default LandingPage;