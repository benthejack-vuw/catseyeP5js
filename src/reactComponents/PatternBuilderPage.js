import React from 'react'

const PatternBuilderPage = () => (
    <div>
        Pattern Builder
    </div>
)

export default PatternBuilderPage;