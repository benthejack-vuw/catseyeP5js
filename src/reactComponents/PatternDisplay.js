import { DomUtils } from 'bj-utils'
import React from 'react'
import { QDOrthoGLCanvas } from '../webGL/qdOrthoGLCanvas'
import GLMirrorPolygon from '../webGL/GLMirrorPolygon'

export class PatternDisplay extends React.Component{
    //props.maxTileSize
    //props.scale
    //props.polyGrid
    //props.image
    //props.showGrid
    //props.textureCoordinates

    _glCanvas;
    _glCtx;

    constructor(props){
        super(props);

        this._glCanvas = document.createElement("canvas");
		this._glCtx = this._glCanvas.getContext("webgl");
	    if (!this._glCtx) {
	      throw("webgl unavailable");
	    }
    }

    render = ()=>(
        <div style={
                {
                width:'60vw',
                height:'60vh',
                float:'left'
                }
            }
        >
            <canvas id='pattern-canvas' width='1000' height='1000'/>
            {this.props.showGrid && <canvas id='grid-canvas'/>}
        </div>
    )
    
    componentDidUpdate = () => {
        this.renderPattern(document.getElementById('pattern-canvas'));
    }

    renderPattern = async (canvas)=>{
        
        const glCanvas = await this.buildGLCanvas(this._glCanvas, this._glCtx);

        this.props.polyGrid.edge_normalized_polygons.forEach(
            (poly) => {
                this.renderPolygon(new GLMirrorPolygon(poly), glCanvas);
            }
        )
        
        this.patternFill(glCanvas.canvas, canvas);

        glCanvas.destroy();
    }

    renderGrid = (e)=>{
        const canvas = e.target;
    }

    buildGLCanvas = async (canvas, glContext)=>{
        const glCanvas = new  QDOrthoGLCanvas(canvas, glContext, vertexShader, fragmentShader, 'u_matrix');
        const image = await DomUtils.buildImageFromURL(this.props.image);
        glCanvas.setView(this.tileSize(), this.renderModelView());
        glCanvas.setTexture(image, 'u_texture');
        glCanvas.glContext.clear(glContext.COLOR_BUFFER_BIT);
        return glCanvas;
    }

    renderPolygon = (glPoly, glCanvas)=>{
        const pts = glPoly.GLPoints();
        glCanvas.setbuffer(pts, 3, "a_position");
        const texcoords = glPoly.GLTexCoords(this.props.textureCoordinates);
        glCanvas.setbuffer(texcoords, 2, 'a_texcoord');

        glCanvas.render(glCanvas.glContext.TRIANGLE_FAN, glPoly.length+2);
    }

    patternFill = (patternCanvas, canvas) => {
        const ctx = canvas.getContext('2d');
        ctx.clearRect(0, 0, canvas.width, canvas.height);
        var pattern = ctx.createPattern(patternCanvas, "repeat");
        ctx.rect(0, 0, canvas.width, canvas.height);
        ctx.fillStyle = pattern;
        ctx.fill();
    }
    
    tileSize = ()=>{
       return {x: this.props.maxTileSize * this.props.scale * this.props.polyGrid.edge_normalized_clipRect.width,
               y: this.props.maxTileSize * this.props.scale * this.props.polyGrid.edge_normalized_clipRect.height};
    }

    renderModelView = () => {
        return {x: this.props.polyGrid.edge_normalized_clipRect.width,
                y: this.props.polyGrid.edge_normalized_clipRect.height}
    }
}

var vertexShader = `
attribute vec2 a_texcoord;
attribute vec4 a_position;

uniform mat4 u_matrix;

varying vec2 v_texcoord;

void main() {
  // Multiply the position by the matrix.
  gl_Position = u_matrix * a_position;

  // Pass the texcoord to the fragment shader.
  v_texcoord = a_texcoord;
}`

var fragmentShader = `
precision mediump float;

// Passed in from the vertex shader.
varying vec2 v_texcoord;

// The texture.
uniform sampler2D u_texture;

void main() {
   gl_FragColor = texture2D(u_texture, v_texcoord);
}`


export default PatternDisplay;