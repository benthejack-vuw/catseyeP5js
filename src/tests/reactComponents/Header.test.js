import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../reactComponents/Header';

test('should render header correctly', ()=>{
    const wrapper = shallow(<Header startLogout={()=>{}}/>);
    expect(wrapper).toMatchSnapshot();
})