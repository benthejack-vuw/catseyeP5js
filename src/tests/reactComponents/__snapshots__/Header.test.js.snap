// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render header correctly 1`] = `
<div
  className="header"
>
  <div
    className="container"
  >
    <header
      className="header__layout"
    >
      <Link
        className="header__home-link"
        to="/"
      >
        <h1>
          Catseye
        </h1>
      </Link>
      <Link
        className="header__help-link"
        to="/help"
      >
        <h2>
          Help
        </h2>
      </Link>
    </header>
  </div>
</div>
`;
